### PR TITLE
[Merged by Bors] - feat(CategoryTheory/SmallObject/Iteration): the unique morphism (limit case)

### PR DESCRIPTION
**Pull request by joelriou**
---

- [x] depends on: #18115
- [x] depends on: #18127

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/SmallObject/Iteration/Basic.lean [hunk 1]
```diff
@@ -18,8 +18,8 @@ a functor `F : { i // i ≤ j } ⥤ C ⥤ C` equipped with the data
 which makes it the `i`th-iteration of `Φ` for all `i` such that `i ≤ j`.
 Under suitable assumptions on `C`, we shall show that this category
 `Iteration ε j` is equivalent to the punctual category (TODO).
-We shall study morphisms in this category, showing first that
-there is at most one morphism between two morphisms (done), and secondly,
+In this file, we show that the there is at most one morphism between
+two objects. In `SmallObject.Iteration.UniqueHom`, we shall show
 that there does always exist a unique morphism between
 two objects (TODO). Then, we shall show the existence of
```
Comment (alreadydone) [inline Mathlib/CategoryTheory/SmallObject/Iteration/Basic.lean:24]:
```suggestion
two objects. Then, we shall show the existence of
```

File: Mathlib/CategoryTheory/SmallObject/Iteration/UniqueHom.lean [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+variable {i : J} {iter₁ iter₂ : Iteration ε (wellOrderSucc i)}
+  (hi : i < wellOrderSucc i) (φ : iter₁.trunc hi.le ⟶ iter₂.trunc hi.le)
+
+/-- Auxiliary definition for `mkOfSucc`. -/
+noncomputable def mkOfSuccNatTransApp (k : J) (hk : k ≤ wellOrderSucc i) :
+    iter₁.F.obj ⟨k, hk⟩ ⟶ iter₂.F.obj ⟨k, hk⟩ :=
+  if hk' : k = wellOrderSucc i then
+    eqToHom (by subst hk'; rfl) ≫ (iter₁.isoSucc i hi).hom ≫
+      whiskerRight (φ.natTrans.app ⟨i, by rfl⟩) _ ≫ (iter₂.isoSucc i hi).inv ≫
+      eqToHom (by subst hk'; rfl)
```
Comment (alreadydone) [inline Mathlib/CategoryTheory/SmallObject/Iteration/UniqueHom.lean:56]: To extend data from `Iio` to `Iic`, you may want to use `piSplitLE` that I introduced [here](https://github.com/leanprover-community/mathlib4/blob/bea07173607044e35d179c483c9e46dba61ea9b6/Mathlib/Order/DirectedInverseSystem.lean#L123-L124).

**Comment by joelriou** [comment]
Thanks @alreadydone for the suggestions!

File: Mathlib/CategoryTheory/SmallObject/Iteration/UniqueHom.lean [hunk 3]
```diff
@@ -136,8 +134,114 @@ noncomputable def mkOfSucc
...
+
+/-- The canonical isomorphism between two objects in the category `Iteration ε j`. -/
+noncomputable def iso : iter₁ ≅ iter₂ where
+  hom := default
+  inv := default
+
+@[simp]
+lemma iso_refl : iso iter₁ iter₁ = Iso.refl _ := by aesop_cat
+
+lemma iso_trans : iso iter₁ iter₂ ≪≫ iso iter₂ iter₃ = iso iter₁ iter₃ := by aesop_cat
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/SmallObject/Iteration/UniqueHom.lean:243]: I don't know how the place where you're applying `iso_trans` looks like, but would it make sense to provide `Unique (iter₁ ≅ iter₂)` instead (or in addition)?

**Review by TwoFX — APPROVED** [review]
Looks good to me, just one question.
bors d+

File: Mathlib/CategoryTheory/SmallObject/Iteration/UniqueHom.lean (hunk 3 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/SmallObject/Iteration/UniqueHom.lean:243]: Thanks for the suggestion. I will try to see if this idea may help or not in the future downsteam PRs.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**